### ISSUE TITLE
Inverting a boolean should use eor

### ISSUE THREAD

**Issue by xfix** [llvm-mos/llvm-mos#279]
The following code:

```c
#include <stdbool.h>
bool invert(bool value) {
    return !value;
}
```

Generates the following assembly:

```
invert:                                 ; @invert
        and     #1
        beq     .LBB0_2
        lda     #0
        rts
.LBB0_2:
        lda     #1
        rts
```

This can be optimized to the following.

```
invert:
    eor #1
    rts
```

**Comment by johnwbyrd** [comment]
Your optimization incorrectly assumes that true == 0x1.

**Comment by mysterymath** [comment]
Yep, fair criticism. I ran this through godbolt, looks like it's coming down from the frontend as a 1-bit XOR:
```
# Machine code for function invert: IsSSA, TracksLiveness

bb.1.entry:
  liveins: $a
  %1:_(s8) = COPY $a
  %2:_(s8) = G_ASSERT_ZEXT %1:_, 1
  %0:_(s1) = G_TRUNC %2:_(s8)
  %3:_(s1) = G_CONSTANT i1 true
  %4:_(s1) = G_XOR %0:_, %3:_; example.c:3:12
  %5:_(s8) = G_ZEXT %4:_(s1); example.c:3:5
  $a = COPY %5:_(s8); example.c:3:5
  RTS implicit $a; example.c:3:5

# End machine code for function invert.
```

Our legalizer is lowering it to the following select:

```
# Machine code for function invert: IsSSA, TracksLiveness, Legalized

bb.1.entry:
  liveins: $a
  %1:_(s8) = COPY $a
  %2:_(s8) = G_ASSERT_ZEXT %1:_, 1
  %0:_(s1) = G_TRUNC %2:_(s8)
  %3:_(s1) = G_CONSTANT i1 true
  %6:_(s1) = G_CONSTANT i1 false
  %4:_(s1) = G_SELECT %0:_(s1), %6:_, %3:_; example.c:3:12
  %7:_(s8) = G_CONSTANT i8 1
  %8:_(s8) = G_CONSTANT i8 0
  %5:_(s8) = G_SELECT %4:_(s1), %7:_, %8:_; example.c:3:5
  $a = COPY %5:_(s8); example.c:3:5
  RTS implicit $a; example.c:3:5

# End machine code for function invert.
```

This is because 1-bit XORs are just how the IR happens to represent negation of boolean flag values, and it's usually better to lower those to control flow than to widen to 8 bits, XOR, then truncate.

But, in this case, the 1-bit XOR is immediately zero-extended (ZEXT) to 8 bits, so we might as well just compute an 8-bit XOR. Should be a relatively simple thing to fix.

**Comment by mysterymath** [comment]
Fixed by 2b1d155b62d63b3a393dc64eb78475bad059dda7

**Comment by mysterymath** [comment]
Had to revert the fix to this, as it causes a pretty severe regression elsewhere.

This will take some doing; the game here is surprisingly tricky, and I don't fully understand all the behavior of all the pieces yet.
